### PR TITLE
[Adamc/restructure-cleanup-codebase] Refactored entire codebase

### PR DESCRIPTION
**Pull request by adamseth2**
## List of Changes

- :white_check_mark: Automatically launches on port 8000 when `npm run dev`
- :white_check_mark: Deleted unused files
- :white_check_mark: Container below Navbar is based on the parent
- :white_check_mark: Standardized 2x3 grid system and showed example of 2x2 container under the maps tab
- :white_check_mark: Added folders for components and renamed some folders
- :white_check_mark: General cleanup of code
  - Fixed all ESLint errors
  - TypeScript errors are only in telemetry & errors
- :white_check_mark: Implemented Pinia, global state management system
  - Removes dependency of importing `Ros` object directly
  - Types are checked rather than using `provide` and `inject` that don't
  - Clear global state
  - Many other benefits but the very long explanation
- :white_check_mark: Refactored usage of roslib and have generic pub and sub to use and replaced all usage with generic ones
- :white_check_mark: Added comments throughout the entire codebase
- :white_check_mark: Added global style buttons under `global.scss`
- :white_check_mark: Updated all npm packages
- :x: Fixed telemetry & other files to use TypeScript
- :x: Added unsubscribe functionality

## Structure of Codebase

- **lib**: hold script and helper functions that could go into .vue files but the code is too long (feel when necessary)
- **store**: hold globals states and anything dealing with ros should be in here

## Questions

1. Right now, the vue components directly use `const rosjs = useRoslibStore();` to access their generic pub and sub. Should .vue use this or should, for example, `Controller.vue` solely rely on the `controllerStore` to interact with ros? This would add a lot more boilerplate, but potentially reduces decoupling and perhaps fewer errors as you have to explicitly create functions that serve one purpose.
2. Any other glaring issues?

### CODE REVIEW

File: src/components/controller/Gamepad.vue [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+onUnmounted(() => {
+  window.removeEventListener('gamepadconnected', onGamePadConnectsHandler);
+  window.removeEventListener('gamepadconnected', onGamePadDisconnectsHandler);
+});
+
+function onGamePadConnectsHandler(e: GamepadEvent) {
+  console.log('Controller connected with index %d\n' + e.gamepad.id, e.gamepad.index);
+  indexToControllerName.set(e.gamepad.index, e.gamepad.id);
+  const state = new ControllerState(
+    'src\\lib\\controller\\drivebaseAndArmBinding.json',
```
Comment (uellenberg) [inline src/components/controller/Gamepad.vue:28]: This isn't necessarily a problem with this PR, just something that caught my eye and might be worth discussing. Do you know why we're using backslashes here? I need to test it, but I wonder if it might cause incompatibilities between platforms.

File: src/lib/interface/robotInfo.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ class RobotInfo {
   constructor(givenROS) {
     this.myROS = givenROS;
 
-    if (this.myROS != null) {
+    if (this.myROS !== null) {
```
Comment (uellenberg) [inline src/lib/interface/robotInfo.ts:10]: Unless I'm misunderstanding this code, `!= null` is actually a valid check here, because it covers both `null` and `undefined`. Generally `!==` and `===` are what you should use, except in the case of `null` if you don't care about whether something is `null` or `undefined`, just that it doesn't have a value.

So, any `!== null` and `=== null` should be replaced with their corresponding single equals varieties when we don't care about whether the value is `null` or `undefined`, just that it doesn't exist. I won't comment on any other times this shows up, so it's probably a good idea to run a full-code search.

File: src/main.ts [hunk 3]
```diff
@@ -29,4 +30,16 @@ const router = createRouter({
   routes,
 });
 
-createApp(App).use(router).mount('#app');
+const app = createApp(App);
+app.config.warnHandler = (msg, vm, trace) => {
+  // We are using logical components, so ignoring this warning
+  if (msg.includes('Component is missing template or render function')) {
```
Comment (uellenberg) [inline src/main.ts:36]: Is there a way to configure Vue to suppress this warning? Relying on the message is liable to having it come back when we upgrade Vue if someone decides to change the wording.

File: src/store/rosTypes.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+type messageType =
```
Comment (uellenberg) [inline src/store/rosTypes.ts:1]: This should be `MessageType` to follow TypeScript naming conventions.

File: src/store/useExample.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  function publishExampleData() {
+    console.log(TOPIC_NAME, TOPIC_TYPE, data);
+    roslib.publish(TOPIC_NAME, TOPIC_TYPE, data.value + '!');
+  }
+  function unsubscribe() {
+    roslib.unsubscribe(TOPIC_NAME, TOPIC_TYPE);
+  }
+  // Return all state, getters and functions
+  return { data, allCapsData, publishExampleData, unsubscribe };
+});
```
Comment (uellenberg) [inline src/store/useExample.ts:27]: Do we need this file?

**Review by uellenberg — CHANGES_REQUESTED** [review]
This looks awesome! Thank you for going through the codebase and doing this. I've left a few comments on things that caught my eye, but first, I noticed that the formatting changes have come back to haunt us, so I did a bit of wranging on the commit history that should fix it.

First, here's the commit ID at the time of review: `a881c89aa20fc4f2d326eff58fb686ba98c61cb7`. As long as we have this, the following can be undone.

I made an `adamc/restructure-cleanup-codebase-2` branch that gets rid of the bad commits. You can switch to it with the following:
```bash
git fetch --all
# Ensure that the branch exists locally
git checkout adamc/restructure-cleanup-codebase-2
git checkout adamc/restructure-cleanup-codebase
git reset --hard adamc/restructure-cleanup-codebase-2
git push --force
```

---

The other thing I noticed is that we still have the patch for roslibjs. As @loukylor pointed out, the generic needs to be provided for the topic class, not the subscribe method. We should remove that patch and modify topics to match.

Speaking of which, where is the code for the topics located? I can't find the code for camera subscribing anymore.

File: src/lib/interface/robotInfo.ts (hunk 2 above)
Comment (adamseth2) [reply src/lib/interface/robotInfo.ts:10]: Nice catch. This robot info needs to actually be refactored entirely. One of the main reasons I chose Pinia is that the useRosStore and the roslib object state itself is always defined, aka the check is not actually necessary.

File: src/store/rosTypes.ts (hunk 4 above)
Comment (adamseth2) [reply src/store/rosTypes.ts:1]: Good catch!

File: src/components/camera/CameraFeed.vue [hunk 6]
```diff
@@ -1,3 +1,4 @@
+<!-- LEGACY CODE NOT USED ANYMORE -->
```
[the remark this answers is not in the document]
Comment (adamseth2) [reply src/components/camera/CameraFeed.vue:1]: When I merged, it accidentally got added again. Deleting again.

File: src/components/camera/CameraFeedPreview.vue [hunk 7]
```diff
@@ -1,3 +1,4 @@
+<!-- LEGACY CODE NOT USED ANYMORE -->
```
[the remark this answers is not in the document]
Comment (adamseth2) [reply src/components/camera/CameraFeedPreview.vue:1]: When I merged, it accidentally got added again. Deleting again.

File: src/store/useRoslib.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    }, 100);
+  }
+  /**
+   * Generic Subscriber to interact with Ros
+   * @param topicName should start with '/' along with topic name
+   * @param messageType Ros Message Type
+   * @param defaultValue? optional starting value
+   * @returns Vue Ref that can be undefined if not provided default value. Can use "as Ref<T>" if default Value is provided
+   */
+  function subscribe<T>(
```
Comment (loukylor) [inline src/store/useRoslib.ts:88]: sub and pub (and anything else that needs to make ros objects) should probably be cached rather than creating a new one every time, especially for frequent publishes.

File: src/pages/Telemetry.vue [hunk 9]
```diff
@@ -106,50 +106,59 @@ function getMoteusStateProxy(param, dataCallback) {
...
 
 <style lang="scss" scoped>
 .page {
   display: flex;
   flex-wrap: wrap;
-  padding: 5px;
-  margin: 5px;
-  justify-content: center;
+  // padding: 5px;
+  // margin: 5px;
```
Comment (loukylor) [inline src/pages/Telemetry.vue:160]: avoid commenting things out without more context, e.g. include a comment at the top saying "these lines aren't needed cuz xx reason"

File: src/pages/Telemetry.vue [hunk 10]
```diff
@@ -106,50 +106,59 @@ function getMoteusStateProxy(param, dataCallback) {
...
+          :class="{ 'button-on': !recordingAll, 'button-off': recordingAll }"
+          @click="recordAllPressed()"
+        >
+          Record all
+        </button>
+      </div>
+      <div class="period-input-container">
+        <b class="period-text">Update Delay in ms</b>
+        <input
+          v-model="update_time_ms"
```
Comment (loukylor) [inline src/pages/Telemetry.vue:122]: naming convention inconsistent (`update_time_ms` is in snake case)

Comment (loukylor) [reply src/pages/Telemetry.vue:122]: same with `update_ms` on the telemetry component

File: src/pages/Telemetry.vue [hunk 11]
```diff
@@ -75,11 +75,11 @@ let robotInfo;
...
 onMounted(() => initialize());
 
-let myRos = inject<Ros>('ros');
+const roslib = useRoslibStore();
 
 function initialize() {
-  if (myRos != null) {
-    robotInfo = new RobotInfo(myRos);
+  if (roslib.ros !== null) {
+    robotInfo = new RobotInfo(roslib.ros);
```
Comment (loukylor) [inline src/pages/Telemetry.vue:82]: wouldn't it make more sense to pass in the roslib object over the ros object?

File: src/components/telemetry/moteus/GenericMotorTelemetry.vue [hunk 12]
```diff
@@ -79,10 +79,10 @@ function dataCallback(result) {
   let json = JSON.parse(result.json_payload);
 
   for (const key in json) {
-    if (json.hasOwnProperty(key)) {
+    if (key in json) {
```
Comment (loukylor) [inline src/components/telemetry/moteus/GenericMotorTelemetry.vue:82]: be careful with the `in` keyword. it will match prototype keys, so for parsed json objects, `"toString" in json`, `"valueOf" in json`, `"isPropertyOf" in json`, etc. will all return true. I personally would change it back to `hasOwnProperty` or `Object.hasOwn`.

File: src/lib/controller/controllerState.ts [hunk 13]
```diff
@@ -1,4 +1,4 @@
-import controllerPub from '../../roslib/controllerPub';
+import { useRoslibStore } from '@/store/useRoslib';
```
Comment (loukylor) [inline src/lib/controller/controllerState.ts:1]: i cant comment on the line, but be consistent whether you want export on the same line as the declaring `export class Foo {}` or on a new line `class Foo; export { Foo }`

File: src/components/controller/Gamepad.vue (hunk 1 above)
Comment (loukylor) [reply src/components/controller/Gamepad.vue:28]: it's because they're using fetch to get the json file, which you ideally shouldn't be doing. vite has built in functionality to import json files: https://vite.dev/guide/features#json. you can pass them into controller state as an object or import them dynamically using glob import: https://vite.dev/guide/features#glob-import

File: src/assets/global.scss [hunk 14]
```diff
@@ -41,8 +42,37 @@ body {
   text-rendering: optimizeLegibility;
   -webkit-font-smoothing: antialiased;
   -moz-osx-font-smoothing: grayscale;
+  height: 100%;
+}
+.button-on {
```
Comment (loukylor) [inline src/assets/global.scss:47]: use scss features; lots of the attributes for button-on and off are the same and therefore redundant

File: src/components/controller/Gamepad.vue (hunk 1 above)
Comment (uellenberg) [reply src/components/controller/Gamepad.vue:28]: Gotcha, I'll address this in my controller changes then.

**Comment by loukylor** [comment]
As for your question Adam, I don't think relying on more stores is really necessary. I'm not super familiar with the state management system you've added, but just from my gut feeling, there doesn't need to be a store. Vue objects are already decently stateful, and global states are needed when many components pull from the same data source or rely on each other's state. Since there isn't a ton of that in the individual components, I don't think it's needed. At some point in the future, there may be a necessity for them, but that isn't now.

File: src/pages/Telemetry.vue (hunk 9 above)
Comment (adamseth2) [reply src/pages/Telemetry.vue:160]: 👍

File: src/lib/interface/robotInfo.ts (hunk 2 above)
Comment (adamseth2) [reply src/lib/interface/robotInfo.ts:10]: Noted for the future.

File: src/components/controller/Gamepad.vue (hunk 1 above)
Comment (adamseth2) [reply src/components/controller/Gamepad.vue:28]: Thanks for mentioning this Kyler. Yeah please get rid of fetch, and yeah just import it directly. I wouldn't use glob as the json files are static and not going to change, but yeah in the future will do this.

File: src/store/useRoslib.ts (hunk 8 above)
Comment (loukylor) [reply src/store/useRoslib.ts:88]: for reference, here is the topic source code: https://github.com/RobotWebTools/roslibjs/blob/develop/src/core/Topic.js

you can see that it is heavily oop based, so the object really should be created and used multiple times. This is unlike the `Message` class that can be thrown away once created

File: src/components/controller/Gamepad.vue (hunk 1 above)
Comment (adamseth2) [reply src/components/controller/Gamepad.vue:28]: Yeah I think this is where we cannot fully don't know what the driver(s) want(s). Yes, there will be different control configurations like tank controls and etc. Hm yeah I guess its the way you want to code it up as the benefit of static is not worrying about async stuff. Thinking of it now, glob import makes the most sense as it actually would be very useful if it is dynamic. A cool feature would have each json file having metadata key-values (aka like 2 keys at the top) that specifies the title like "tank control" & "Bob's Standard Control" and perhaps color.  The title and color would then be used in the navbar bar to display what control mode it is in. I know Vlad mentioned that the driver should be able to change the controls while the code is running, but I think that is very low priority as that would involve either storing it on local storage (then another computer wouldn't work) or automatically saving the json which the amount of time and effort isn't worth it. I think the easiest thing is if the driver really wants their own controls, they can copy the json file and change the values themselves, which isn't very hard and they shouldn't need to do it a lot.

File: src/pages/Telemetry.vue (hunk 11 above)
Comment (adamseth2) [reply src/pages/Telemetry.vue:82]: This check is actually not needed as roslib and roslib.ros should always not be null or undefined thanks to pinia(was just a quick fix). I will refactor Telemetry in another PQ.

File: src/lib/controller/controllerState.ts (hunk 13 above)
Comment (adamseth2) [reply src/lib/controller/controllerState.ts:1]: Changed all classes to `export default Foo` 👍

File: src/main.ts (hunk 3 above)
Comment (adamseth2) [reply src/main.ts:36]: Yeah I don't like this solution either, but I googled and honestly, this is the only way I could find. If you find a way, then feel free to fix it. Hopefully, in future Vue versions, there will be a way

File: src/store/useRoslib.ts (hunk 8 above)
Comment (adamseth2) [reply src/store/useRoslib.ts:88]: Thanks for linking to the source code. This actually helps a separate problem I had with unsubscribe as from my understanding it needs to be the same Topic object to be able to unsub. I'll experiment with it when I implement unsubscribe, but I am not 100% sure this would actually impact performance.

TLDR: each sub needs a topic object to be able to unsubscribe due to how they coded it up. For Pubs, I will look into if caching helps when I implement unsub

File: src/components/telemetry/moteus/GenericMotorTelemetry.vue (hunk 12 above)
Comment (adamseth2) [reply src/components/telemetry/moteus/GenericMotorTelemetry.vue]: image
@uellenberg I think the eslint and formatted changed this automatically. To note in the future is make sure to the formatter version is the same in tsconfig.json, so it doesn't change it for some reason. I had to add the following in tsconfig.json 
 ```json   
  // Added to allow Object.hasOwn(). We are targeting the latest versions of javascript.
    "target": "ES2023",
    "lib": ["ES2022"]
 ```

 json.hasOwnProperty gives eslint so I think that is why it was automatically changed, so we will be using Object.hasOwn:
`  82:14  error    Do not access Object.prototype method 'hasOwnProperty' from target object                     no-prototype-builtins`

Comment (adamseth2) [reply src/components/telemetry/moteus/GenericMotorTelemetry.vue]: Tbh though, hopefully nobody uses prototypes ever... But changed it to keep it clean

**Comment by adamseth2** [comment]
@uellenberg  The camera code will be shipped later as the rover code on the camera side completely changed. And as the camera code here is outdated already, thought it doesn't make sense to still have it.
Will look into whether patch is needed still. I tried inserting the type into Topic but there was a problem with undefined, but I haven't looked into it too much.

File: src/components/telemetry/moteus/GenericMotorTelemetry.vue (hunk 12 above)
Comment (loukylor) [reply src/components/telemetry/moteus/GenericMotorTelemetry.vue]: I did some research into that ESLint error, and it seems that it is valid in that some objects may not have the member function `hasOwnProperty`, so you may want to switch to `Object.hasOwn` 

see: https://stackoverflow.com/questions/39282873/object-hasownproperty-yields-the-eslint-no-prototype-builtins-error-how-to

File: src/lib/controller/controllerState.ts (hunk 13 above)
Comment (loukylor) [reply src/lib/controller/controllerState.ts:1]: Let me clarify, cuz I still see a few instances of this in the code. Some code uses the single line syntax to export, seen here: https://github.com/TrickfireRobotics/mission-control/blob/625c3f2ebc8ef0f8cccedd6e6eebb21dbd919235/src/roslib/heartbeatPub.ts#L6

Some places define what they want to export, then export it on a separate line, like so: https://github.com/TrickfireRobotics/mission-control/blob/625c3f2ebc8ef0f8cccedd6e6eebb21dbd919235/src/script/saveCSVData.ts#L58

You should choose one and stick to it, for consistency.

File: src/main.ts (hunk 3 above)
Comment (uellenberg) [reply src/main.ts]: Can we add a dummy render function? In React, you can return null for it to do nothing. I would prefer to not globally disable warnings for a local usage.

File: src/components/telemetry/moteus/GenericMotorTelemetry.vue (hunk 12 above)
Comment (uellenberg) [reply src/components/telemetry/moteus/GenericMotorTelemetry.vue]: > I did some research into that ESLint error, and it seems that it is valid in that some objects may not have the member function `hasOwnProperty`, so you may want to switch to `Object.hasOwn`
> 
> see: https://stackoverflow.com/questions/39282873/object-hasownproperty-yields-the-eslint-no-prototype-builtins-error-how-to

The problem with hasOwnProperty (at least, from what I remember) is that it's susceptible to prototype pollution, as it exists on the element itself (rather than a global like Object). I don't think it's a huge issue, but if we're worrying about prototype pollution anyway, then we should worry about this as well.

I'm not sure that this is an actual issue though, as (from what I remember) a `for ... in` loop only goes through the enumerable properties of an object - of which prototypes are not (perhaps by default). We should be able to eliminate the if statement entirely, unless there's a specific reason for it (in which case, it should be in a comment).

Comment (loukylor) [reply src/components/telemetry/moteus/GenericMotorTelemetry.vue]: I didn't even see it was in a for in loop, you're probably right that this if check isn't required at all. The mdn docs have a [helpful table[(https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties#querying_object_properties) that implies for in is actually more restrictive than `hasOwn` or `hasOwnProperty` so I'm leaning towards removing the loop altogether and just making a note of this in later code.

**Comment by adamseth2** [comment]
Okay, I now remember why I did the whole patch thing last year and did not pass into Topic. Unless I am missing a step, the problem is with passing the type into ROSLIB.Topic is that the response would expect T. This would work though but I don't know why Roslib is coded this way but the subscribe callback actually **ALWAYS** returns the json Type `{data : unknown}` not T (I've done extensive testing and experimenting with the rover).

This code is from me playing in the master branch, but this is what typescript would yell at you for:

Passing type into ROSLIB.Topic:
image

Passing the exact json type object that could remedy this problem but also this is a base case feature that every topic should have:
image


Okay. I've thought about it again and I'll remove the patch because long term it is a bad solution when updating dependencies and that could potentially break the codebase. I'll just make it a type, and make it generic. Side effect it would have to be something like the 2nd image of passing in a wrapper Type (it would be generic just too lazy to update the screenshot)

**Comment by uellenberg** [comment]
Gotcha, that makes sense. It looks like Topic takes in the type of the entire message, including the data field. Presumably there are messages without a data field, but I'm not sure what they are. The easiest solution would probably making an:
```ts
export interface Msg<T> {
    data: T,
}
```

Somewhere, then writing `Topic<Msg<number>>`. That's a bit annoying though, but I'm not sure if data can ever not be there, so I'm hesitant to modify the types to always include a data field.

**Comment by loukylor** [comment]
If you look at the [typing for roslibjs](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/roslib/index.d.ts#L494), it shows that the parameter in the callback of subscribe is of type `TMessage`, which is the type parameter that you pass into topic. So, in the code that is erroring, typescript think that `message` has the type `number`. It doesn't have a `.data` attribute, so of course it errors.

As how to actually fix the code, you have two options. Again, if you look at the [typing](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/roslib/index.d.ts#L451) again, notice that `TMessage = Message`, i.e. it when no type parameter is passed, `TMessage` will equal `Message` by default. Message has no attributes, so we probably shoudn't use it, but this is option one.

Option two, which is probably what we should do, is to pass in our own types into topic. Standard messages have a `data` field, so you can make a generic `StdMsg<T>` type, but all others that we define will need their own type and may not have a `data` field. These types that we define and aren't standard messages will need to be coordinated between the rover teams to ensure we're interpreting the data correctly.

**Comment by adamseth2** [comment]
Yeah, I have implemented option 2. Ohhh I never fully realized that is why `data` was always there (2 images below shows both containing `data`. Thanks for pointing that out.

I'll keep note of it, moving on forward. I was debating if I should pull out the entire Typescript power and all the conditional typing stuff, but right now it's like 20x easier if there is a custom type, they ensure they have `.data` attribute and that's only if the mission control needs it anyway. The other stuff attributes like header etc like in CompressedImage don't really matter as we should only care about the actual data. (though the way it is coded right now it could be implemented in the future not too hard just not a huge priority as we aren't using much custom types )

image


image

**Comment by adamseth2** [comment]
Ready for another look. I believe it is basically merge-ready now.

File: src/store/useRoslib.ts [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+        ros.connect(serverHost);
+      }
+    }, 100);
+  }
+  function createTopic<T>(
+    topicName: string,
+    topicType: TopicType,
+    compression: RosCompressionType = 'cbor',
+  ) {
+    return new ROSLIB.Topic<StdMsg<T>>({
```
Comment (loukylor) [inline src/store/useRoslib.ts:82]: note that not all msg types will have a data field; it would be better to pass `T` into `Topic` and make the caller use the `StdMsg<T>` type.

File: src/store/useRoslib.ts [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+   * @param options.isDebugging? optional prints to console for debugging
+   * @returns a callback function that publishes given Data
+   */
+  function createPublisher<T extends TopicType>(options: {
+    topicName: string;
+    topicType: T;
+    isDebugging?: boolean;
+  }) {
+    const { topicName, topicType, isDebugging } = options;
+    const topic = new ROSLIB.Topic({
```
Comment (loukylor) [inline src/store/useRoslib.ts:169]: this was brought up before, but it would be best to cache topic objects

**Review by loukylor — DISMISSED** [review]

**Comment by loukylor** [comment]
just a couple small changes are needed

**Comment by adamseth2** [comment]
Fixed all changes

**Review by uellenberg — APPROVED** [review]
Looks good to me, thank you! As this is a large change, please make sure to use a merge commit.

**Comment by loukylor** [comment]
Sounds good, I'm quite busy as of now and won't be able to really deep dive for another few days. Since this is time sensitive waiting that long isn't too viable. Thanks Jonah for reviewing and Adam for doing all this. Good work everyone!

**Comment by adamseth2** [comment]
No worries Kyler! I thought you were busy with other stuff and you already did so much. Again thanks to both of you for getting me many rounds of iterative feedback they were extremely helpful. Hopefully, this is the last BIG PR, and the rest would be very small. Again thanks!
